### PR TITLE
[MRG] Edit BiDAF to accept multiple choice format input

### PR DESCRIPTION
**Pull request by nelson-liu**
This PR edits our MC BiDAF model to actually take multiple choice data as input.

TODO:
- [x] edit `model.get_feed_dict` to process multiple choice input
  - This was pretty trivial to do, as I operated under the assumption that the options data would be treated the same way as the passage data, since they have the same shape.
- [x] the `squad/prepro` script normally reads the squad data and serializes tokenized versions to a JSON file. We thus need to create a new version of this script to read in our MC data and serialize it in a way that is compatible with the current model code (so try to mimic the original output as much as possible, but add some fields for the options)
- [x] The `read_data.update_config` method needs to be updated to add values for `config.max_num_options` and `config.max_option_size`.
- [x] Right now, there's a custom class (`basic/evaluator.MultiGPUF1Evaluator`) to do evaluation and return an F1/EM score. We need to change this from {take the output probabilities, get the best span, and compute an F1/EM} to {take the output probabilities, get the argmax for the most likely option, write that output to a file or something}. We can then do a post-hoc analysis of the accuracy; i ultimately think it's limiting to have it all done in this class. The current behavior of the `test` routine is to "evaluate", and I think it would be better as just "predict" given some input.

### CODE REVIEW

**Comment by nelson-liu** [comment]
@mattg, @pdasigi, note that this code actually does not work for training a vanilla DA BiDAF model --- the sole function is to do MC prediction. It probably would not be hard to readd that functionality in the future, but I just wanted to get MC prediction working first.

Here are some instructions on how to run (`bi-att-flow` is where you cloned this repo):
0) Make sure you have the proper BiDAF dependencies as listed in `requirements.txt`. Note that this code uses tensorflow 0.11, you can find install instructions [here](https://www.tensorflow.org/versions/r0.11/get_started/os_setup).
1) Copy the BiDAF pretrained model (the `basic` directory) from `/efs/data/dlfa/models/vanilla_pretrained_bidaf_for_mc` to `bi-att-flow/out/`. Just copy the whole folder over, so the path looks like `bi-att-flow/out/basic/`
2) Figure out which questions + background files you want to use, and a descriptive `outname` for naming the output file. As an example, i'll pick:
  - background file = `/efs/data/dlfa/questions/intermediate/processed/intermediate_4_questions_with_lucene_background/intermediate_4_dev_background.tsv`
  - questions file = `/efs/data/dlfa/questions/intermediate/processed/intermediate_4_dev/question_and_answer/questions.tsv`
  - out name = `intermediate_4_dev_lucene`

3) Now, run the data preprocessing script (still under `squad/prepro.py`, I want to refactor this eventually...) with the following format: `python -m squad.prepro $backgroundpath $questionpath $outname`. It should dump some output files with the `outname` prefix to `bi-att-flow/data/squad/`

4) Now that we have the data files prepared, we can run inference. To do so, run the command `python -m basic.cli --out_name=<outname> --batch_size=32`. Thus, for the example above we'd run `python -m basic.cli --out_name=intermediate_4_dev_lucene --batch_size=32`. I get OOM errors when running with the default batch size of 60, so i've tuned it down to 32.

If all goes well, you should see a progress bar as it makes predictions. at the end, it'll print an accuracy value. It'd be great if one of you could check that you're able to run the code on your machine.

File: basic/main.py [hunk 1]
```diff
@@ -128,11 +128,14 @@ def _train(config):
 
 
 def _test(config):
-    test_data = read_data(config, 'test', True)
+    assert config.out_name
```
Comment (nelson-liu) [inline basic/main.py:131]: this just verifies that `config.out_name` is set, i should probably add a more informative error message

File: basic/model.py [hunk 2]
```diff
@@ -157,7 +157,7 @@ def _build_forward(self):
                         word_emb_mat = tf.get_variable("word_emb_mat", dtype='float', shape=[word_vocab_size, word_emb_size],
                                                        initializer=get_initializer(config.emb_mat))
                     else:
-                        word_emb_mat = tf.get_variable("word_emb_mat", shape=[word_vocab_size, ], dtype='float')
+                        word_emb_mat = tf.get_variable("word_emb_mat", shape=[word_vocab_size, word_emb_size], dtype='float')
```
Comment (nelson-liu) [inline basic/model.py:160]: not sure how the graph was even compiling with this bug, but...

File: basic/model.py [hunk 3]
```diff
@@ -209,7 +209,7 @@ def _build_forward(self):
         passage_len = tf.reduce_sum(tf.cast(self.passage_mask, 'int32'), 2)
         # [batch_size]
         question_len = tf.reduce_sum(tf.cast(self.question_mask, 'int32'), 1)
-        # [batch_size, num_sentences]
+        # [batch_size, num_options]
```
Comment (nelson-liu) [inline basic/model.py:212]: note that `options_mask` is of shape `[batch_size, num_options, option_size]`. This sums it up so we have a matrix of `[batch_size, num_options]` indicating the length of each option. I hackily use this as a mask for the option logits later down the line (by converting the length matrix of to a bool matrix representing whether we have an option there, or whether it's all padding)

File: basic/model.py [hunk 4]
```diff
@@ -220,17 +220,17 @@ def _build_forward(self):
             encoded_question = tf.concat(2, [fw_outputs, bw_outputs])
             if config.share_lstm_weights:
                 tf.get_variable_scope().reuse_variables()
-                # [batch_size, num_sentences, sentence_size, 2*hidden_size]
+                # [batch_size, num_sentences, sentence_size, 2*hidden_size] (verified)
```
Comment (nelson-liu) [inline basic/model.py:223]: i manually checked the shapes of these as a sanity check, i should remove these comments...

File: basic/model.py [hunk 5]
```diff
@@ -327,8 +328,7 @@ def _build_forward(self):
             flattened_passage_mask = tf.reshape(self.passage_mask,
                                                 [-1, num_sentences * sentence_size])
             # shape: [-1, num_sentences*sentence_size, 2xhidden_dim]
-            weighted_passage = envelope * flattened_encoded_passage
-
+            weighted_passage = tf.tile(tf.expand_dims(envelope, 2), [1, 1, hidden_size*2]) * flattened_encoded_passage
```
Comment (nelson-liu) [inline basic/model.py:331]: Oddly enough, this wouldn't broadcast.

**Review by matt-gardner — APPROVED** [review]
